pydevices/acq400Devices/acq400_base.py

```diff
@@ -311,12 +423,147 @@ def store(self):
+        decims = uut.read_decims()
+        dt = 1 / ((round(float(uut.s0.SIG_CLK_MB_FREQ.split(" ")[1]), -4)) * 1e-9)
+        tb_ns = self._create_time_base(decims, dt)
+
+        self.DECIMS.putData(decims)
+        self.DT.putData(dt)
+        self.TB_NS.putData(tb_ns)
+
+    def store(self):
+        uut = acq400_hapi.Acq400(self.node.data())
+        self.chans = []
+        nchans = uut.nchan()
+        for ii in range(nchans):
+            self.chans.append(getattr(self, 'INPUT_%3.3d'%(ii+1)))
+
+        uut.fetch_all_calibration()
+        eslo = uut.cal_eslo[1:]
+        eoff = uut.cal_eoff[1:]
+        channel_data = uut.read_channels()
+
+        for ic, ch in enumerate(self.chans):
+            if ch.on:
+                ch.putData(channel_data[ic])
+                ch.EOFF.putData(float(eoff[ic]))
+                ch.ESLO.putData(float(eslo[ic]))
+                expr = "{} * {} + {}".format(ch, ch.ESLO, ch.EOFF)
+                ch.CAL_INPUT.putData(MDSplus.Data.compile(expr))
+
+	    self.create_time_base(uut)
+        # return None
     STORE=store
 
 
+    def arm():
+        # A customised ARM function for the acq2106_MR setup.
+        uut = acq400_hapi.Acq400(self.node.data())
+        shot_controller = acq400_hapi.ShotController(uut)
+        shot_controller.run_shot(remote_trigger=self.selects_trg_src(uut, self.trg0_src.data()))
+        return None
+    ARM = arm
+
+
+    def selects_trg_src(self, uut, src):
+        def select_trg_src():
+            uut.s0.SIG_SRC_TRG_0 = src
+        return select_trg_src
+
+
+    def denormalise_stl(self, stl):
+
+        # lines = args.stl.splitlines()
+        lines = stl.splitlines()
+        stl_literal_lines = []
+        for line in lines:
+            if line.startswith('#') or len(line) < 2:
+                # if args.verbose:
+                #     print(line)
+                continue
+            else:
+                action = line.split('#')[0]
+
+                if action.startswith('+'): # support relative increments
+                    delayp = '+'
+                    action  = action[1:]
+                else:
+                    delayp = ''
+
+                delay, state = [int(x) for x in action.split(',')]
+                # delayk = int(delay * self.Fclk.data() / 1000000)
+                delayk = int(delay * 40000000 / 1000000)
+                delaym = delayk - delayk % self.MR10DEC.data()
+                state = state << self.evsel0.data()
+                elem = "{}{:d},{:02x}".format(delayp, delaym, state)
+                stl_literal_lines.append(elem)
+                # if args.verbose:
+                #     print(line)
+
+        return "\n".join(stl_literal_lines)
+
+
+    def init(self):
+        # args.uuts = [ acq400_hapi.Acq2106(u, has_comms=False) for u in args.uut ]
+        uut = acq400_hapi.Acq2106(self.node.data())
+        # master = args.uuts[0]
+        self.STL.putData("/home/dt100/PROJECTS/acq400_hapi/user_apps/STL/acq2106_test10.stl")
```

Comment:
Hardcoded path?

Reply:
This is something I noticed and I wanted to bring to their attention. I'm not sure why this was done. I wonder if they were just testing. We are not using Multi Rate devices yet. But for sure I need to remind them about this.